es6/animations.js

```diff
@@ -5,180 +5,309 @@ import {
         }
+    }
+    throw `computeTheOther got (${start}, ${delta}, ${end}) as args and didn't know what to do`
+}
 
-        if (start_state === undefined)
-            start_state = end_state - amt
-        if (end_state === undefined)
-            end_state = start_state + amt
-        if (amt === undefined)
-            amt = end_state - start_state
+const exactlyOneIsUndefined = (val1, val2) => {
+    return ((val1 === undefined || val2 === undefined)
+            && !(val1 === undefined && val2 === undefined))
+}
 
-        if (start_state + amt != end_state) {
-            console.log({start_state, end_state, amt})
-            throw 'Conflicting values, Animation end_state != start + amt'
-        }
+const isNumber = (val) => {
+    return typeof(val) === 'number'
+}
+
+const applyDefaultsAndValidateTimes = (start_time, duration, end_time) => {
+    if (start_time === undefined) start_time = Date.now()
 
-        return {start_state, end_state, amt}
+    if (duration === undefined && end_time === undefined) {
+        duration = 1000; // removing this semi-colon results in a gnarly parse error
+        [duration, end_time] = computeTheOther(start_time, duration, end_time)
+    } else if (exactlyOneIsUndefined(duration, end_time)) {
+        [duration, end_time] = computeTheOther(start_time, duration, end_time)
     } else {
-        if (start_state === undefined) start_state = {}
-        if (end_state === undefined) end_state = {}
-        if (amt === undefined) amt = {}
+        throw 'only one of (duration, end_time) should be passed in, not both.'
+    }
 
-        if (key !== undefined)
-            return checked_animation_amt({start_state: start_state[key], end_state: end_state[key], amt: amt[key]})
+    if (start_time > end_time) {
+        throw `start_time (${start_time}) > end_time (${end_time})`
+    }
+    return [start_time, duration, end_time]
+}
 
-        if (typeof(start_state) !== 'object' || typeof(end_state) !== 'object' || typeof(amt) !== 'object') {
-            throw 'Incompatible types passed as {start_state, end_state, amt}, must all be objects or numbers'
-        }
 
-        let keys = Object.keys(start_state)
-        if (!keys.length) keys = Object.keys(end_state)
-        if (!keys.length) keys = Object.keys(amt)
+const validateAnimation = (animation) => {
+    const end_time = animation.end_time
+    const end_state = animation.end_state
 
-        return keys.reduce((acc, key) => {
-            const amts = checked_animation_amt({start_state: start_state[key], end_state: end_state[key], amt: amt[key]})
-            acc.start_state[key] = amts.start_state
-            acc.end_state[key] = amts.end_state
-            acc.amt[key] = amts.amt
-            return acc
-        }, {start_state: {}, end_state: {}, amt: {}})
+    const computed_end_state = computeAnimatedState({
+        animations:[animation],
+        warped_time:_end_time
```

Comment:
I think this `_end_time` should be just `end_time` in this function, right?

Reply:
nice catch